### PR TITLE
Fix optimize label scan (2)

### PR DESCRIPTION
**Pull request by raz-mon**
Fix Label-Scan optimization to:
1. Take into account valid swap options (for instance mandatory ones for `MATCH` clauses, see tests to further understand).
2. Work for `OPTIONAL MATCH` clauses as well as mandatory `MATCH` clauses (no mixing between the clauses!).

Solves [MOD-5093](https://redislabs.atlassian.net/browse/MOD-5093)

### CODE REVIEW

File: tests/flow/test_optimizations_plan.py [hunk 1]
```diff
@@ -475,4 +475,66 @@ def test29_optimize_label_scan_cached_label_id(self):
...
+        # no nodes of label Z, but it is optional, so the second most tempting
+        # label (Q) must be traversed first (order swapped with N)
+        query = """MATCH (n:N) MATCH (n:Q) OPTIONAL MATCH (n:Z) RETURN n"""
+        plan = graph.execution_plan(query)
+        self.env.assertIn("Node By Label Scan | (n:Q)", plan)
+        self.env.assertIn("Conditional Traverse | (n:N)->(n:N)", plan)
+
+        # assert correctness of the results
+        res = graph.query(query)
+        self.env.assertEquals(res.result_set[0][0], Node(label=['N', 'Q'], properties={'v': 2}))
```
Comment (swilly22) [inline tests/flow/test_optimizations_plan.py:513]: assert number of results, expecting 1 row

File: tests/flow/test_optimizations_plan.py [hunk 2]
```diff
@@ -475,4 +475,66 @@ def test29_optimize_label_scan_cached_label_id(self):
...
+
+        # make sure that labels from different `OPTIONAL MATCH` clauses are not
+        # "mixed" in Label-Scan optimization
+        query = "OPTIONAL MATCH (n:N) OPTIONAL MATCH (n:M) RETURN n"
+        plan = graph.execution_plan(query)
+
+        # make sure `N` is the first label traversed, even though there are less
+        # labels with label `M`
+        self.env.assertIn("Node By Label Scan | (n:N)", plan)
+        self.env.assertIn("Conditional Traverse | (n:M)->(n:M)", plan)
```
Comment (swilly22) [inline tests/flow/test_optimizations_plan.py:540]: Missing a test that verifies that the correct label (one with minimum entities) is selected when used in `OPTIONAL MATCH` similar to `test30` above but when specifying `OPTIONAL`

File: src/execution_plan/ops/shared/scan_functions.h [hunk 3]
```diff
@@ -8,16 +8,18 @@
 
 // Storage struct for label data in node and index scans.
 typedef struct {
-	const char *alias;   // Alias of the node being traversed.
-	const char *label;   // Label of the node being traversed.
-	int label_id;        // Label ID of the node being traversed.
+	const char *alias;  // alias of the node being traversed
+	const char *label;  // label of the node being traversed
+	int label_id;       // label ID of the node being traversed
+	int *candidates;   // candidates for swapping in Label-Scan optimization
```
Comment (swilly22) [inline src/execution_plan/ops/shared/scan_functions.h:14]: `int` should be replaced with `LabelID` for both `candidates` and `label_id`
Also make sure `//` are all aligned.

Lastly should this become: `const int *candidates` ?

File: src/execution_plan/execution_plan_build/build_match_op_tree.c [hunk 4]
```diff
@@ -82,8 +82,16 @@ static void _ExecutionPlan_ProcessQueryGraph
 			Schema *s = GraphContext_GetSchema(gc, label, SCHEMA_NODE);
 			if(s != NULL) label_id = Schema_GetID(s);
 
+			// build candidates for swap in Label-Scan optimization
+			uint *candidates = array_new(uint, 0);
+			for(uint i = 0; i < label_count; i++) {
+				uint _label_id = QGNode_GetLabelID(src, i);
+				if(_label_id != label_id) {
+					array_append(candidates, _label_id);
+				}
```
Comment (swilly22) [inline src/execution_plan/execution_plan_build/build_match_op_tree.c:91]: I think it is perfectly fine to include `label_id` within the `candidates` list.
Also I think it would be OK to use a stack based array: `uint *candidates[label_count]` moreover maybe we can share the `QGNode` labels array pointer ?

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: tests/flow/test_optimizations_plan.py (hunk 2 above)
Comment (raz-mon) [reply tests/flow/test_optimizations_plan.py:540]: See line 530

File: src/execution_plan/execution_plan_build/build_match_op_tree.c (hunk 4 above)
Comment (raz-mon) [reply src/execution_plan/execution_plan_build/build_match_op_tree.c:91]: Given that the amount of labels is not expected to be more than ~10 (average is less), I don't think this overhead is significant. I think it's a bit clearer this way.. Do you think the overhead is worth the clarity here?

File: src/execution_plan/ops/op_node_by_label_scan.c [hunk 5]
```diff
@@ -226,5 +226,9 @@ static void NodeByLabelScanFree(OpBase *op) {
 		UnsignedRange_Free(nodeByLabelScan->id_range);
 		nodeByLabelScan->id_range = NULL;
 	}
-}
 
+	if(nodeByLabelScan->n.n != NULL) {
```
Comment (swilly22) [inline src/execution_plan/ops/op_node_by_label_scan.c:230]: `if(nodeByLabelScan->n != NULL)`

File: src/execution_plan/ops/op_node_by_label_scan.c [hunk 6]
```diff
@@ -226,5 +226,9 @@ static void NodeByLabelScanFree(OpBase *op) {
...
+		NODE_CTX_FREE(nodeByLabelScan->n);
+		nodeByLabelScan->n.n = NULL;
```
Comment (swilly22) [inline src/execution_plan/ops/op_node_by_label_scan.c:232]: `nodeByLabelScan->n = NULL;`

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/execution_plan/ops/shared/scan_functions.c [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    char *alias,       // alias
+    char *label,       // label
+    LabelID label_id,  // label id
+    QGNode *n          // node
+) {
+    NodeScanCtx *ctx = rm_malloc(sizeof(NodeScanCtx));
+    ctx->alias = alias;
+    ctx->label = label;
+    ctx->label_id = label_id;
+    ctx->n = n;
```
Comment (swilly22) [inline src/execution_plan/ops/shared/scan_functions.c:23]: `ctx->n = QGNode_Clone(n);`

File: src/execution_plan/ops/shared/scan_functions.c [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    return ctx;
+}
+
+// clones a context
+NodeScanCtx *NodeScanCtx_Clone
+(
+    NodeScanCtx *ctx  // context
+) {
+    NodeScanCtx *clone = rm_malloc(sizeof(NodeScanCtx));
+    clone->alias = ctx->alias;
```
Comment (swilly22) [inline src/execution_plan/ops/shared/scan_functions.c:34]:
```c
memcpy(clone, clone, sizeof(NodeScanCtx));
clone->n = QGNode_Clone(ctx->n);
```

File: src/execution_plan/ops/shared/scan_functions.c [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+    return clone;
+}
+
+// frees a context
+void NodeScanCtx_Free
+(
+    NodeScanCtx *ctx  // context
+) {
+    if(ctx->n != NULL) {
```
Comment (swilly22) [inline src/execution_plan/ops/shared/scan_functions.c:47]:
```c
ASSERT(ctx != NULL);
ASSERT(ctx->n != NULL);
```

File: src/execution_plan/ops/shared/scan_functions.c [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        QGNode_Free(ctx->n);
+        ctx->n = NULL;
```
Comment (swilly22) [inline src/execution_plan/ops/shared/scan_functions.c:49]: Redundant, next line is going to free `ctx`

File: src/execution_plan/ops/shared/scan_functions.c [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    rm_free(ctx);
+}
```
Comment (swilly22) [inline src/execution_plan/ops/shared/scan_functions.c:53]: file should end with an empty line

File: src/execution_plan/ops/shared/scan_functions.h [hunk 12]
```diff
@@ -16,20 +17,23 @@ typedef struct {
...
+(
+    NodeScanCtx *ctx  // context
+);
 
-#define NODE_CTX_FREE(_ctx)	(QGNode_Free(_ctx.n))
+// frees a context
+void NodeScanCtx_Free
+(
+    NodeScanCtx *ctx  // context
+);
```
Comment (swilly22) [inline src/execution_plan/ops/shared/scan_functions.h:39]: file should end with an empty line

File: src/execution_plan/ops/shared/scan_functions.h [hunk 13]
```diff
@@ -16,20 +17,23 @@ typedef struct {
...
-#define NODE_CTX_CLONE(_ctx)       \
-({                                 \
-	NodeScanCtx ctx = _ctx;        \
-	ctx.n = QGNode_Clone(_ctx.n);  \
-	ctx;                           \
-})
+// clones a context
+NodeScanCtx *NodeScanCtx_Clone
+(
+    NodeScanCtx *ctx  // context
```
Comment (swilly22) [inline src/execution_plan/ops/shared/scan_functions.h:32]: `const NodeScanCtx *ctx`

File: src/execution_plan/ops/shared/scan_functions.h [hunk 14]
```diff
@@ -16,20 +17,23 @@ typedef struct {
...
-	.label_id = (_label_id),                         \
-	.n = (_n)                                        \
-}
+// allocates and returns a new context
+NodeScanCtx *NodeScanCtx_New
+(
+    char *alias,       // alias
+    char *label,       // label
+    LabelID label_id,  // label id
+    QGNode *n          // node
```
Comment (swilly22) [inline src/execution_plan/ops/shared/scan_functions.h:26]: `const QGNode *n`

File: src/execution_plan/optimizations/utilize_indices.c [hunk 15]
```diff
@@ -377,26 +377,27 @@ void reduce_scan_op
...
-		scan->n.label_id = min_label_id;
+		scan->n->label = min_label_str;
+		scan->n->label_id = min_label_id;
 	}
 
 	FT_FilterNode *root = _Concat_Filters(filters);
-	OpBase *indexOp = NewIndexScanOp(scan->op.plan, scan->g, scan->n, rs_idx,
+	NodeScanCtx *ctx_clone = NodeScanCtx_Clone(scan->n);
+	OpBase *indexOp = NewIndexScanOp(scan->op.plan, scan->g, ctx_clone, rs_idx,
 			root);
```
Comment (swilly22) [inline src/execution_plan/optimizations/utilize_indices.c:401]: Maybe ?
```c
OpBase *indexOp = NewIndexScanOp(scan->op.plan, scan->g, scan->n, rs_idx, root);
scan->n = NULL;
```

File: src/execution_plan/ops/op_node_by_label_scan.c [hunk 16]
```diff
@@ -207,7 +211,8 @@ static OpResult NodeByLabelScanReset(OpBase *ctx) {
 static OpBase *NodeByLabelScanClone(const ExecutionPlan *plan, const OpBase *opBase) {
 	ASSERT(opBase->type == OPType_NODE_BY_LABEL_SCAN);
 	NodeByLabelScan *op = (NodeByLabelScan *)opBase;
-	OpBase *clone = NewNodeByLabelScanOp(plan, NODE_CTX_CLONE(op->n));
+	NodeScanCtx *ctx_clone = NodeScanCtx_Clone(op->n);
+	OpBase *clone = NewNodeByLabelScanOp(plan, ctx_clone);
```
Comment (swilly22) [inline src/execution_plan/ops/op_node_by_label_scan.c:215]: One line? (less than 80 chars?)
```c
return NewNodeByLabelScanOp(plan, NodeScanCtx_Clone(op->n));
```

File: src/execution_plan/execution_plan_build/build_match_op_tree.c [hunk 17]
```diff
@@ -84,7 +84,7 @@ static void _ExecutionPlan_ProcessQueryGraph
 
 			// resolve source node by performing label scan
 			QGNode *clone = QGNode_Clone(src);
-			NodeScanCtx ctx = NODE_CTX_NEW(alias, label, label_id, clone);
+			NodeScanCtx *ctx = NodeScanCtx_New((char *)alias, (char *)label, label_id, clone);
```
Comment (swilly22) [inline src/execution_plan/execution_plan_build/build_match_op_tree.c:87]: Didn't we wanted `NodeScanCtx_New` to perform the clone internally ?
`NodeScanCtx *ctx = NodeScanCtx_New((char *)alias, (char *)label, label_id, src);`

**Review by swilly22 — CHANGES_REQUESTED** [review]
Looks good, a few more comments

File: src/execution_plan/ops/shared/scan_functions.c (hunk 11 above)
Comment (raz-mon) [reply src/execution_plan/ops/shared/scan_functions.c:53]: It does..

File: src/execution_plan/ops/shared/scan_functions.h (hunk 12 above)
Comment (raz-mon) [reply src/execution_plan/ops/shared/scan_functions.h:39]: Also, it does.

**Review by swilly22 — APPROVED** [review]
